### ISSUE TITLE
Visual Studio 2017 crashes when trying to add custom python environment

### ISSUE THREAD

**Issue by Arrowana** [Microsoft/PTVS#3432]
In Tools -> Python -> Python Environments -> + Custom

After adding the python path prefix, auto-detect works fine.
However, clicking on "Apply" crashes Visual Studio 2017, without any relevant error message.

Visual Studio Enterprise 2017 15.4.5
Python 3.6.3 64 bit

**Comment by huguesv** [comment]
I wasn't able to reproduce this locally.

Could you please save the information from Tools, Python, Diagnostic Info and post it here or email it to ptvshelp@microsoft.com. (no need to include analysis log)

There's a good chance that it will include a call stack for the devenv crash.

**Comment by huguesv** [comment]
Thanks for the log email.

The relevant part:
```
Parameter name: root ---> System.UriFormatException: Invalid URI: The format of the URI could not be determined.
   at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)
   at System.Uri..ctor(String uriString, UriKind uriKind)
   at Microsoft.PythonTools.Infrastructure.PathUtils.MakeUri(String path, Boolean isDirectory, UriKind kind, String throwParameterName)
   --- End of inner exception stack trace ---
   at Microsoft.PythonTools.Infrastructure.PathUtils.MakeUri(String path, Boolean isDirectory, UriKind kind, String throwParameterName)
   at Microsoft.PythonTools.Infrastructure.PathUtils.GetAbsoluteFilePath(String root, String relativePath)
   at Microsoft.PythonTools.Interpreter.PythonRegistrySearch.TryReadConfiguration(String company, String tag, RegistryKey tagKey, RegistryKey installKey, Boolean pythonCoreCompatibility, InterpreterArchitecture assumedArch)
   at Microsoft.PythonTools.Interpreter.PythonRegistrySearch.Search(RegistryKey root, InterpreterArchitecture assumedArch)
   at Microsoft.PythonTools.Interpreter.CPythonInterpreterFactoryProvider.DiscoverInterpreterFactories()
   at Microsoft.PythonTools.Interpreter.CPythonInterpreterFactoryProvider.EnsureInitialized()
   at Microsoft.PythonTools.Interpreter.CPythonInterpreterFactoryProvider.GetInterpreterConfigurations()
   at Microsoft.PythonTools.Interpreter.InterpreterRegistryService.GetConfigurations()
   at Microsoft.PythonTools.Interpreter.InterpreterRegistryService.get_Configurations()
   at Microsoft.PythonTools.Intellisense.OutOfProcProjectAnalyzer.<Initialize>d__26.MoveNext()

```

So we wrote some invalid data in the registry. Not sure what caused it, but it may help to know what's in there currently.

For comparison, here's my entries for my custom added Python 3.6:
```
Windows Registry Editor Version 5.00

[HKEY_CURRENT_USER\Software\Python\VisualStudio]
"DisplayName"="Custom environment"

[HKEY_CURRENT_USER\Software\Python\VisualStudio\Python36]
"SysArchitecture"="64bit"
"SysVersion"="3.6"
"DisplayName"="Python36"

[HKEY_CURRENT_USER\Software\Python\VisualStudio\Python36\InstallPath]
@="C:\\Program Files\\Python36"
"WindowedExecutablePath"="C:\\Program Files\\Python36\\pythonw.exe"
"ExecutablePath"="C:\\Program Files\\Python36\\python.exe"

```

If you can, please post your entries (use regedit and click export on the HKCU\Software\Python\VisualStudio tree node).

**Comment by Arrowana** [comment]
My registry:

Windows Registry Editor Version 5.00

[HKEY_CURRENT_USER\Software\Python\VisualStudio]
"DisplayName"="Custom environment"

[HKEY_CURRENT_USER\Software\Python\VisualStudio\Python36]
"SysArchitecture"="64bit"
"SysVersion"="3.6"
"DisplayName"="Python36"

[HKEY_CURRENT_USER\Software\Python\VisualStudio\Python36\InstallPath]
@="C:\\Program Files\\Python36"
"WindowedExecutablePath"="C:\\Program Files\\Python36\\pythonw.exe"
"ExecutablePath"="C:\\Program Files\\Python36\\python.exe"

So exactly what you have.

**Comment by huguesv** [comment]
Looking at the call to GetAbsoluteFilePath from the error call stack, this is only called for PythonCore tag, not for custom interpreters. I don't think it's the cause of the crash you are seeing when adding a custom interpreter.

Looking at the difference in the code between 15.5 and 15.6, we've already added a catch for the case where GetAbsoluteFilePath fails due to bad input (coming from InstallPath registry value). It's probably worth looking at all your PythonCore/(tag)/InstallPath (Default) value to see if there's one that's not an absolute path. In case that exception is what causes later issues.

If you can make it crash again when adding a custom interpreter, you should open Event Viewer, and look at the error entries (red icon) under Windows Logs/Application. There should be one for devdiv.exe, hopefully with useful details. That's what I was hoping to find in the logs you sent.

**Comment by Arrowana** [comment]
I can find this in the Error viewer:
```
Faulting application name: devenv.exe, version: 15.0.27130.0, time stamp: 0x5a20cf11
Faulting module name: KERNELBASE.dll, version: 10.0.15063.726, time stamp: 0x32774c02
Exception code: 0xe0434352
Fault offset: 0x000ecbb2
Faulting process id: 0x83f8
Faulting application start time: 0x01d36e39415fcfe6
Faulting application path: C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\devenv.exe
Faulting module path: C:\WINDOWS\System32\KERNELBASE.dll
Report Id: de0fc866-75ea-4e47-a0f8-b7aa85485537
Faulting package full name: 
Faulting package-relative application ID: 
```

Not a lot of additional info

**Comment by zooba** [comment]
Alternatively, if you collect the info from Tools, Python, Diagnostic Info, we should have the stack trace in there (if we collected one) from any recent crashes. You can either export and upload it here, or just copy the error info out for us.

**Comment by huguesv** [comment]
He's done that already, look for the email from Pierre Duval.

**Comment by Arrowana** [comment]
Still broken after update to 15.5.2, any update?

**Comment by zooba** [comment]
There's a fix in for 15.6.

Without knowing what the invalid path is or where it came from, there's nothing else we can do here.
